os/arch/arm/src/armv7-m/up_assert.c

```diff
@@ -250,95 +251,103 @@ static void up_dumpstate(void)
 {
 	struct tcb_s *rtcb = this_task();
 	uint32_t sp = up_getsp();
-	uint32_t ustackbase;
-	uint32_t ustacksize;
+	uint32_t stackbase = 0;
+	uint32_t stacksize = 0;
 #if CONFIG_ARCH_INTERRUPTSTACK > 3
-	uint32_t istackbase;
-	uint32_t istacksize;
+	uint32_t istackbase = 0;
+	uint32_t istacksize = 0;
 #endif
 
-	/* Get the limits on the user stack memory */
-
-	ustackbase = (uint32_t)rtcb->adj_stack_ptr;
-	ustacksize = (uint32_t)rtcb->adj_stack_size;
+	/* Get the limits for each type of stack */
 
+	stackbase = (uint32_t)rtcb->adj_stack_ptr;
+	stacksize = (uint32_t)rtcb->adj_stack_size;
 #if CONFIG_ARCH_INTERRUPTSTACK > 3
-	/* Get the limits on the interrupt stack memory */
-
 	istackbase = (uint32_t)&g_intstackbase;
 	istacksize = (CONFIG_ARCH_INTERRUPTSTACK & ~3);
-
-	/* Show interrupt stack info */
-
-	lldbg("sp:     %08x\n", sp);
-	lldbg("IRQ stack:\n");
-	lldbg("  base: %08x\n", istackbase);
-	lldbg("  size: %08x\n", istacksize);
-#ifdef CONFIG_STACK_COLORATION
-	lldbg("  used: %08x\n", up_check_intstack());
 #endif
+	bool isIrqAssert = false;
+	bool isSpCorrupt = false;
 
-	/* Does the current stack pointer lie within the interrupt
-	 * stack?
+	/* Check if the assert location is in user thread or IRQ handler.
+	 * If the irq_num is lesser than NVIC_IRQ_USAGEFAULT, then it is
+	 * a fault and not an irq.
 	 */
-
-	if (sp <= istackbase && sp > istackbase - istacksize) {
-		/* Yes.. dump the interrupt stack */
-
-		up_stackdump(sp, istackbase);
+	if (g_irq_nums[2] || (g_irq_nums[1] && g_irq_nums[0] > NVIC_IRQ_USAGEFAULT)) {
+		/* Assert in nested irq */
+		isIrqAssert = true;
+		lldbg("Code asserted in nested IRQ state!\n");
+	} else if ((g_irq_nums[1] && (g_irq_nums[0] <= NVIC_IRQ_USAGEFAULT)) || (g_irq_nums[0] > NVIC_IRQ_USAGEFAULT)) {
+		/* Assert in irq */
+		isIrqAssert = true;
+		lldbg("Code asserted in IRQ state!\n");
+	} else {
+		/* Assert in user thread */
+		lldbg("Code asserted in normal thread!\n");
+		if (current_regs) {
+			/* If assert is in user thread, but current_regs is not NULL,
+			 * it means that assert happened due to a fault. So, we want to
+			 * reset the sp to the value just before the fault happened
+			 */
+			sp = current_regs[REG_R13];
+		}
 	}
 
-	/* Extract the user stack pointer if we are in an interrupt handler.
-	 * If we are not in an interrupt handler.  Then sp is the user stack
-	 * pointer (and the above range check should have failed).
-	 */
-
-	if (current_regs) {
-		sp = current_regs[REG_R13];
-		lldbg("sp:     %08x\n", sp);
-	}
+	/* Print IRQ handler details if required */
 
-	lldbg("User stack:\n");
-	lldbg("  base: %08x\n", ustackbase);
-	lldbg("  size: %08x\n", ustacksize);
-#ifdef CONFIG_STACK_COLORATION
-	lldbg("  used: %08x\n", up_check_tcbstack(rtcb));
+	if (isIrqAssert) {
+		uint8_t crashIrqNum = 1;
```

Comment:
Let us not check this here. 
Please define a variable called irqnum at the beginning and set the variable to proper value when we are determining whether its a assert in irq or user thread in the above conditionals at line 276.